frontend/views/pages/Contributions.vue

```diff
@@ -208,28 +187,79 @@ export default {
       }
     }
   },
-  validations: {
-    form: {
-      incomeAmount: {
-        [L('field is required')]: required,
-        [L('cannot be negative')]: v => v >= 0,
-        [L('cannot have more than 2 decimals')]: decimals(2)
-      },
-      pledgeAmount: { required, minValue: v => v >= 0, decimals: decimals(2) }
-    }
-  },
   computed: {
     ...mapGetters([
+      'ourUsername',
       'groupSettings',
+      'groupMembersCount',
       'groupProfile',
+      'groupProfiles',
+      'thisMonthsPayments',
       'groupMincomeFormatted',
-      'ourUsername'
+      'groupIncomeDistribution',
+      'globalProfile'
     ]),
-    doesReceiveMonetary () {
-      return !!this.fakeStore.receiving.monetary && !this.ephemeral.isEditingIncome
+    memberGroupProfile () {
+      return this.groupProfile(this.ourUsername) || {}
+    },
+    upTo () {
+      const amount = this.memberGroupProfile[this.memberGroupProfile.incomeDetailsType]
+      if (!amount) return false
+      return this.currency.displayWithCurrency(this.needsIncome ? this.groupSettings.mincomeAmount - amount : amount)
+    },
+    doesReceive () {
+      return this.needsIncome || this.receivingNonMonetary
+    },
+    doesGive () {
+      return this.hasPayments || this.memberGroupProfile.nonMonetaryContributions
+    },
+    needsIncome () {
+      return this.memberGroupProfile.incomeDetailsType === 'incomeAmount'
+    },
+    hasPayments () {
+      return Object.keys(this.givingMonetary.list).length > 0
+    },
+    currency () {
+      return currencies[this.groupSettings.mincomeCurrency]
+    },
+    distribution () {
+      return this.thisMonthsPayments.frozenDistribution || this.groupIncomeDistribution
+    },
+    receivingNonMonetary () {
+      const groupProfiles = this.groupProfiles
+      return Object.keys(groupProfiles)
+        .filter(key => key !== this.ourUsername && groupProfiles[key].nonMonetaryContributions)
+        .reduce((list, username) => {
+          const nonMonetary = groupProfiles[username].nonMonetaryContributions
+          nonMonetary.forEach((what) => {
+            const contributionIndex = list.findIndex(x => x.what === what)
```

Comment:
Instead of having ~3 loops (`reduce`, `forEach`, and `findIndex`), create an empty object at the top of the `receivingNonMonetary` called `services` and use that to check if a service has already been listed, so that there's only one loop

Reply:
*oops, you do still need to loop through `nonMonetary` with `forEach`, but you can get rid of `findIndex` using this method

Reply:
so:

```js
if (services[what]) {
  services[what].push(displayName)
} else {
  services[what] = [displayName]
}
```

Reply:
This can be a TODO to optimize in the future